File: packages/chat-sync/src/task-updates.ts
export type TaskUpdateState = 'pending' | 'failed';

export class TaskUpdates {
  private states: Readonly<Partial<Record<string, TaskUpdateState>>> = {};
  private listeners = new Set<() => void>();

  getSnapshot = () => this.states;
  subscribe = (listener: () => void) => {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  };

  private set(id: string, state?: TaskUpdateState) {
    const next = { ...this.states };
    if (state) next[id] = state;
    else delete next[id];
    this.states = next;
    this.listeners.forEach((listener) => listener());
  }

  async run(id: string, update: () => unknown | Promise<unknown>) {
    if (this.states[id] === 'pending') return;
    this.set(id, 'pending');
    try {
      await update();
      this.set(id);
    } catch {
      this.set(id, 'failed');
    }
  }
}

File: apps/android/core/contract/src/main/kotlin/com/beegreat/contract/TaskUpdates.kt
package com.beegreat.contract

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

enum class TaskUpdateState { Pending, Failed }

/** Mirrors chat-sync's pending lock and retry state without changing saved task data. */
class TaskUpdates {
  private val mutableStates = MutableStateFlow<Map<String, TaskUpdateState>>(emptyMap())
  val states = mutableStates.asStateFlow()

  suspend fun run(id: String, update: suspend () -> Unit) {
    synchronized(this) {
      if (mutableStates.value[id] == TaskUpdateState.Pending) return
      mutableStates.update { it + (id to TaskUpdateState.Pending) }
    }
    try {
      update()
      mutableStates.update { it - id }
    } catch (canceled: CancellationException) {
      mutableStates.update { it - id }
      throw canceled
    } catch (_: Exception) {
      mutableStates.update { it + (id to TaskUpdateState.Failed) }
    }
  }
}
